components/browser/state/src/main/java/mozilla/components/browser/state/reducer/EngineStateReducer.kt

```diff
@@ -0,0 +1,49 @@
+package mozilla.components.browser.state.reducer
+
+import mozilla.components.browser.state.action.EngineAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.EngineState
+import mozilla.components.browser.state.state.SessionState
+
+internal object EngineStateReducer {
+
+    /**
+     * [EngineAction] Reducer function for modifying a specific [EngineState]
+     * of a [SessionState].
+     */
+    fun reduce(state: BrowserState, action: EngineAction): BrowserState = when (action) {
+        is EngineAction.LinkEngineSessionAction -> state.copyWithEngineState(action.sessionId) {
+            it.copy(engineSession = action.engineSession)
+        }
+        is EngineAction.UnlinkEngineSessionAction -> state.copyWithEngineState(action.sessionId) {
+            it.copy(engineSession = null, engineSessionState = null)
+        }
+        is EngineAction.UpdateEngineSessionStateAction -> state.copyWithEngineState(action.sessionId) {
+            it.copy(engineSessionState = action.engineSessionState)
+        }
+    }
+}
+
+private fun BrowserState.copyWithEngineState(
+    tabId: String,
+    update: (EngineState) -> EngineState
+): BrowserState {
+    // Currently we map over both lists (tabs and customTabs). We could optimize this away later on if we know what
+    // type we want to modify.
+    return copy(
+        tabs = tabs.map { current ->
+            if (current.id == tabId) {
+                current.copy(engineState = update.invoke(current.engineState))
+            } else {
+                current
+            }
+        },
+            customTabs = customTabs.map { current ->
+                if (current.id == tabId) {
+                    current.copy(engineState = update.invoke(current.engineState))
+                } else {
+                    current
+                }
+            }
```

Comment:
nit: somehow `tabs` and `customTabs` have a different indentation.

Reply:
fixed!